### PR TITLE
[TE] Rootcause DAG execution framework

### PR DESCRIPTION
**Pull request by apucher**
* DAG based execution framework

    * pipelines as operators

    * entities as data

* basic unit tests

* builds on #1360

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * centralized repository of valid URNs.
+ */
+public class Entity {
+  final String urn;
+  final double score;
+
+  public static final Comparator<Entity> HIGHEST_SCORE_FIRST = new Comparator<Entity>() {
+    @Override
+    public int compare(Entity o1, Entity o2) {
+      return -Double.compare(o1.score, o2.score);
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java]: can we do -1 * Double.compare(o1.score, o2.score) for readability.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * <br/><b>NOTE:</b> a subclass of {@code Entity} may be returned by pipelines in order to pass
+ * along transient meta-data for individual entities.
+ *
+ * <br/><b>NOTE:</b> due to the potentially vast number of nodes in the knowledge graph there is no
+ * centralized repository of valid URNs.
+ */
+public class Entity {
+  final String urn;
+  final double score;
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java]: does this mean entity has to always have a score? 
I thought only a pipeline will attach the score to an entity. An entity by itself should not have a score right?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  /**
+   * Executes the pipeline given the execution context set up by the RCAFramework. Returns entities
+   * as determined relevant given the user-specified search context (contained in the execution
+   * context).
+   *
+   * @param context pipeline execution context
+   * @return pipeline results
+   */
+  public abstract PipelineResult run(PipelineContext context);
+}
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java]: add newline

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * associating search context entities with other relevant entities. Also performs relative ranking
+ * of associated entities in terms of importance to the user.
+ *
+ * @see RCAFramework
+ */
+public abstract class Pipeline {
+  final String name;
+  final Set<String> inputs;
+
+  public Pipeline(String name, Set<String> inputs) {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java]: what are the inputs? entityURNs?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/PipelineContext.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * pipelines.
+ */
+public class PipelineContext {
+  final Map<String, Set<Entity>> inputs;
+
+  public PipelineContext(Map<String, Set<Entity>> inputs) {
+    this.inputs = inputs;
+  }
+
+  public Map<String, Set<Entity>> getInputs() {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/PipelineContext.java]: whats the string in the inputs map, lets name the variable appropriately. inputs gave me a notion that its array or list

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * Container object for framework execution results. Holds the results aggregated across all
+ * pipeline executions as well as the results for each individual pipeline (keyed by pipeline name).
+ *
+ */
+public final class RCAFrameworkResult {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java]: Can we rename it to represent the result of one execution

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  public RCAFrameworkResult(Set<? extends Entity> results, Map<String, PipelineResult> pipelineResults) {
+    this.results = new HashSet<>(results);
+    this.pipelineResults = pipelineResults;
+  }
+
+  public Set<Entity> getResults() {
+    return results;
+  }
+
+  public Map<String, PipelineResult> getPipelineResults() {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java]: string is the name of the pipeline?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/StaticPipeline.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+   * @param inputs pipeline inputs
+   * @param entities entities to emit as result
+   */
+  public StaticPipeline(String name, Set<String> inputs, Set<Entity> entities) {
+    super(name, inputs);
+    this.entities = entities;
+  }
+
+  @Override
+  public PipelineResult run(PipelineContext context) {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/StaticPipeline.java]: where do we use inputs? Do we return the same result without checking the inputs?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java (hunk 2 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java]: I ended up including meta data with the entity instances. In a sense, each instance of Entity is meta data already - the urn itself is only a reference to the "thing" outside of the framework

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java (hunk 1 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Entity.java]: done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java (hunk 3 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java]: done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java (hunk 4 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java]: pipeline names

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/PipelineContext.java (hunk 5 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/PipelineContext.java]: added javadoc

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java (hunk 6 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java]: done

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java (hunk 7 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/RCAFrameworkResult.java]: yes. added javadoc

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/StaticPipeline.java (hunk 8 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/StaticPipeline.java]: The static pipeline always returns the same result, regardless of the input.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+   * Executes the pipeline given the execution context set up by the RCAFramework. Returns entities
+   * as determined relevant given the user-specified search context (contained in the execution
+   * context).
+   *
+   * @param context pipeline execution context
+   * @return pipeline results
+   */
+  public abstract PipelineResult run(PipelineContext context);
+
+}
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java]: add new line

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java (hunk 4 above)
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/Pipeline.java]: please name the variables accordingly. Better yet have a concrete class to represent PipelineName

**Pull request merged**